File: construction_management/construction_management/report/ra_bill_sales_invoice_report/ra_bill_sales_invoice_report.js
frappe.query_reports["RA Bill Sales Invoice Report"] = {
	filters: [
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
		{
			fieldname: "boq",
			label: __("BOQ"),
			fieldtype: "Link",
			options: "BOQ",
		},
		{
			fieldname: "ra_bill",
			label: __("RA Bill"),
			fieldtype: "Link",
			options: "RA Bill",
		},
		{
			fieldname: "sales_invoice",
			label: __("Sales Invoice"),
			fieldtype: "Link",
			options: "Sales Invoice",
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "ra_bill_status",
			label: __("RA Bill Status"),
			fieldtype: "Select",
			options: "\nDraft\nSubmitted\nApproved\nInvoiced\nCancelled",
		},
		{
			fieldname: "invoice_status",
			label: __("Invoice Status"),
			fieldtype: "Select",
			options: "\nDraft\nReturn\nCredit Note Issued\nSubmitted\nPaid\nPartly Paid\nUnpaid\nOverdue\nCancelled",
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
	],
};

File: construction_management/construction_management/report/ra_bill_sales_invoice_report/ra_bill_sales_invoice_report.py
from frappe import _

from construction_management.construction_management.report.report_utils import (
	ensure_report_access,
	first_currency,
	get_invoice_rows_for_ra_bills,
	get_ra_bill_rows,
	parse_filters,
	sum_field,
	summary_metric,
)


def execute(filters=None):
	filters = parse_filters(filters)
	ensure_report_access()
	columns = get_columns(filters)
	data = get_data(filters)
	report_summary = get_report_summary(data)
	return columns, data, None, None, report_summary


def get_columns(filters=None):
	return [
		{
			"label": _("Sales Invoice"),
			"fieldname": "sales_invoice",
			"fieldtype": "Link",
			"options": "Sales Invoice",
			"width": 170,
		},
		{"label": _("RA Bill"), "fieldname": "ra_bill", "fieldtype": "Link", "options": "RA Bill", "width": 160},
		{"label": _("Project"), "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 160},
		{"label": _("BOQ"), "fieldname": "boq", "fieldtype": "Link", "options": "BOQ", "width": 160},
		{"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 160},
		{"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 115},
		{"label": _("Grand Total"), "fieldname": "grand_total", "fieldtype": "Currency", "options": "currency", "width": 140},
		{
			"label": _("Outstanding Amount"),
			"fieldname": "outstanding_amount",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 160,
		},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 120},
	]


def get_data(filters):
	if filters.get("ra_bill_status"):
		filters.status = filters.ra_bill_status

	ra_bills = get_ra_bill_rows(filters, require_invoice=True)
	rows = get_invoice_rows_for_ra_bills(ra_bills)

	if filters.get("invoice_status"):
		rows = [row for row in rows if row.get("status") == filters.invoice_status]

	return rows


def get_report_summary(data):
	currency = first_currency(data)
	return [
		summary_metric("Total Invoiced", sum_field(data, "grand_total"), currency=currency),
		summary_metric("Total Outstanding", sum_field(data, "outstanding_amount"), currency=currency),
	]
